WebConsole: support NULL SQL values for display and xgress

Karakatiza666 opened the pull request:
Is this a user-visible change (yes/no): yes

Fix https://github.com/feldera/feldera/issues/1392 : NULL's show up as NaN's

Fix https://github.com/feldera/feldera/issues/1051 : WebConsole should use JSON row update format to support inserting ARRAY columns

Fix https://github.com/feldera/feldera/issues/1169 : Clicking on 'Got To Next Page" icon in Inspect Table view throws an exception

mihaibudiu commented on web-console/src/lib/components/streaming/import/randomData/ExampleDisplay.tsx:
```diff
@@ -1,29 +1,30 @@
+ *
+ * Also indicates if the example was modified after generation due to type
+ * constraints (overflow etc.).
+ * @param param0
```
Remove

mihaibudiu commented on web-console/src/lib/functions/ddl.ts:
```diff
@@ -5,80 +5,110 @@
 import { match, P } from 'ts-pattern'
 
-import { GridCellParams } from '@mui/x-data-grid-pro'
+export type SQLValueJS = string | number | BigNumber | boolean | Dayjs | SQLValueJS[]
```
no null?

mihaibudiu commented on web-console/src/lib/compositions/streaming/inspection/useQuantiles.ts:
```diff
@@ -18,7 +18,7 @@ function useQuantiles() {
     async (
       egressParams: Arguments<typeof HttpInputOutputService.httpOutput>,
-      setQuantiles: Dispatch<SetStateAction<any[][] | undefined>>,
+      setQuantiles: Dispatch<SetStateAction<(SQLValueJS | null)[][] | undefined>>,
```
It's better to add null to SqlValueJS.

mihaibudiu commented on web-console/src/lib/functions/ddl.ts:
```diff
@@ -216,22 +246,17 @@ export const findBaseType = (type: ColumnType): ColumnType => {
  * @returns
  */
-export const parseValueSafe = (sqlType: ColumnType, value: string) => {
+export const parseCSVValue = (sqlType: ColumnType, value: string) => {
```
why CSV?

Karakatiza666 replied:
This function is used just to parse the value from a CSV row. Soon when I am refactoring to xgress via JSON rather than CSV this will no longer be used, so I decided to mark its purpose appropriately

mihaibudiu commented on web-console/src/lib/components/streaming/inspection/SQLValueDisplay.tsx:
```diff
@@ -0,0 +1,14 @@
+
+import { Typography } from '@mui/material'
+
+export const SQLValueDisplay = ({ value, type }: { value: SQLValueJS; type: ColumnType }) => {
```
why not create a type called TypedSqlValue which has both value and type?

Karakatiza666 replied:
A good idea, but will require quite some refactoring to do consistently which is not a priority

mihaibudiu commented on web-console/src/lib/compositions/streaming/import/useDefaultRows.ts:
```diff
@@ -9,7 +9,7 @@ import { match } from 'ts-pattern'
 import { GridApi } from '@mui/x-data-grid-pro'
 
-export const getDefaultValue = (columntype: ColumnType) =>
+export const getDefaultValue = (columntype: ColumnType): SQLValueJS =>
```
what does this mean?

Karakatiza666 replied:
`: SQLValueJS` states the return type; with it the on-hover function definition tooltip resolves the return type as a concise `: SQLValueJS`, without - an equivalent, but less convenient `string | number | boolean | BigNumber | dayjs.Dayjs`

mihaibudiu commented on web-console/src/lib/functions/ddl.ts:
```diff
@@ -5,80 +5,110 @@
     .with({ type: SqlType.DATE }, () => {
-      return (value: Dayjs | string) => {
+      return value => {
+        invariant(typeof value === 'string' && isDayjs(value))
```
these invariants should not be checked here, but when you construct the value

Karakatiza666 replied:
The code was initially loosely typed, so I am gradually tightening up the types and adding checks that I will be able to drop later on, as I do not want to spend a separate PR going over the values' lifecycle

mihaibudiu commented on web-console/src/lib/services/pipelineManagerQuery.ts:
```diff
@@ -470,6 +476,31 @@ export const mutationDeleteService = (
   }
 })
 
+export const mutationHttpIngressJson = (
```
no idea what goes on here

mihaibudiu commented on web-console/src/lib/components/streaming/inspection/InspectionTable.tsx:
```diff
@@ -315,37 +321,38 @@ const InspectionTableImpl = ({
                 return params.row.record[col.name]
               },
-              renderHeader: () => <SQLTypeHeader col={col}></SQLTypeHeader>
+              // valueFormatter: getValueFormatter(col.columntype),
```
remove?

mihaibudiu commented on web-console/src/lib/components/streaming/import/SQLValueInput.tsx:
```diff
@@ -14,83 +16,94 @@ import { TextField, TextFieldProps } from '@mui/material'
+          ...props
+        }))
+        .with(SqlType.DATE, () => ({
+          type: 'date',
```
can't this just use some tostring method of the type?

Karakatiza666 replied:
Not sure what do you mean? Now the date value is temporarily input as a string value, and is getting converted to a date JS type (via a provided method/function) elsewhere.

Karakatiza666 commented on web-console/src/lib/components/input/BigNumberInput.tsx:
```diff
@@ -1,10 +1,12 @@
+import { StandardTextFieldProps, SxProps, TextField, TextFieldProps, Theme } from '@mui/material'
 
-type EventElementType = Omit<HTMLTextAreaElement | HTMLInputElement, 'value'> & { value?: BigNumber }
+type EventElementType = Omit<HTMLInputElement, 'value'> & { value?: BigNumber | null }
```
?: states that the value can be `undefined` which is distinct from `null`. There could be a use case where either are possible non-exclusively

mihaibudiu replied:
I really hope there is no undefined value anywhere in this code. Null makes things hard enough.

Karakatiza666 replied:
The way BigNumberInput is currently used there is no undefined value state. With proper typing both `undefined` and `null` are useful, easy and quite secure to deal with

gz commented:
I'm getting invariant failed on the insert dialog:

<img width="1217" alt="image" src="https://github.com/feldera/feldera/assets/127654/9ced7aaf-522a-4db7-8150-3a0a05fd86af">


for this SQL

```
CREATE TABLE customer_t (  
    f1 BOOLEAN,
    f2 TINYINT,
    f3 SMALLINT,
    f4 INTEGER,
    f5 BIGINT,
    //f6 FLOAT,
    f7_0 FLOAT32,
    f7 DOUBLE,
    f8 DECIMAL,
    f9 CHAR,
    f10 VARCHAR,
    f11 TEXT, 
    f12 BINARY,
    f13 VARBINARY,
    f14 TIME,
    f15 DATE,
    f16 TIMESTAMP,
    //f17 INTERVAL,
    f18 BOOLEAN ARRAY
    //f19 NULL
); 

```

most likely it's due to some types not supported by insert dialog, but this should in general not fail the UI (it should just not insert anything for them).

gz commented:
I looks like it sorts numbers alphabetically rather than numerical 
<img width="205" alt="image" src="https://github.com/feldera/feldera/assets/127654/b0ef7e4d-38ad-48f0-9492-cc94875565ec">

gz commented:
It would be good to be able to select NULL here as an option in the generator:
<img width="622" alt="image" src="https://github.com/feldera/feldera/assets/127654/a2f3869f-82fa-49d8-ae82-88de86db9179">

gz commented:
I'm unable to set the boolean field to false:

https://github.com/feldera/feldera/assets/127654/51db3036-1365-401f-b676-68be04e14870

Karakatiza666 commented:
Added proper sorting for different SQL types
Fixed bugs with serialization of different SQL types when sending them over API
Refactored CSV format of HTTP ingress/eggress to a JSON format: Fix https://github.com/feldera/feldera/issues/1051

>It would be good to be able to select NULL here as an option in the generator:

I am leaving this feature for a separate issue and PR

> Unable to set this field to null
Fields can be set to NULL by clicking the X icon to the left in the left of the resprctive input

gz commented:
getting invariant failed here:

https://github.com/feldera/feldera/assets/127654/d8c0752d-8e5c-4ec4-8bf8-0bc44b9be14b

gz commented:
browsing to the right of a table leads to invariant failed... I used the SQL program pasted below, I did set the decimal field f8_0 of one of the rows to null
https://github.com/feldera/feldera/assets/127654/51b83496-4565-4786-8d48-85fff2df540b

gz commented:
I'm reiterating a point I made a few weeks ago: if the invariants are so hard to ensure maybe they shouldn't be invariants. Ideally the code should be written in a robust way, e.g., in this case clearly the table library can continue to display things just fine even if some type doesn't match expectations and we don't need to crash the entire UI element.

gz commented:
https://github.com/feldera/feldera/assets/127654/cc049ee5-6eed-479b-8b79-e14a85946a6d

gz commented:
the behavior for char(1) seems weird it doesn't set it to NULL when I click the X only once cell looses focus, it works as expected for varchar with unlimited length
https://github.com/feldera/feldera/assets/127654/e582f3dd-7e6e-49c3-aac3-8ec1ba2d9c4a

gz commented:
invariant failed for array
https://github.com/feldera/feldera/assets/127654/2f00bc4f-b9e7-4152-a87f-99fb7fdc7877

gz commented:
editing array field generally doesnt seem to work:

https://github.com/feldera/feldera/assets/127654/613cb128-28c9-481d-af35-4c543cd38285

gz commented:
can't seem to edit array:
[Screencast from 03-19-2024 05:16:28 PM.webm](https://github.com/feldera/feldera/assets/127654/a5a104de-b5da-4d38-9290-72f4058ba7fc)

gz commented:
should this be allowed? I can't send it afterwards
[Screencast from 03-19-2024 05:30:56 PM.webm](https://github.com/feldera/feldera/assets/127654/4c4632c1-7ae3-4503-8aac-141e52c970c9)

gz commented:
this is an odd one, timestamp array and cursor jumps back whenever I try to delete  a character
[Screencast from 03-19-2024 05:41:53 PM.webm](https://github.com/feldera/feldera/assets/127654/a1b513b9-3808-45c7-8f67-cf36775f09d2)

gz commented:
this is a bit odd because it doesn't show null when I click the X. maybe it's because of the mm/dd/yyyy template?

[Screencast from 03-19-2024 05:44:44 PM.webm](https://github.com/feldera/feldera/assets/127654/7441bf2e-be4d-4620-
b32a-8bf960a3856b)

gz commented:
the DATE field seems to pre-fill the values form the cell in the mm/dd/yyyy template that shows in the editor but the timestamp field doesnt seem to do that?

[Screencast from 03-19-2024 05:46:27 PM.webm](https://github.com/feldera/feldera/assets/127654/e34e63b1-c3a8-44c8-b347-91aa0205b893)

Karakatiza666 commented:
> this is a bit odd because it doesn't show null when I click the X. maybe it's because of the mm/dd/yyyy template?

It is like this for DATE and TIMESTAMP, and there doesn't seem to be a minimalist fix for that, so I suggest leaving this as-is

gz commented:
would be nice if this displays the base type too
![Screenshot from 2024-03-19 17-47-26](https://github.com/feldera/feldera/assets/127654/037d6821-a05b-4d90-adae-0032e1b828b9)

gz commented:
cool works great now. Looking forward to the tests for it :) .. FWIW I also did some X ARRAY ARRAY types and it works fine (aside from when the type is decimal as reported above)!

gz commented:
If table is empty it shouldn't spin here which would indicate to the user there is a problem with loading the data. I think it used to say that the table is empty?

https://github.com/feldera/feldera/assets/127654/e38c4ad5-700c-45e0-bc1d-e86be900abb6

gz commented:
If I set varchar generation to word

<img width="495" alt="image" src="https://github.com/feldera/feldera/assets/127654/eb6e47fe-33c7-45d1-93d4-06a49ed34e34">

all elements are empty strings in the array, seems like there is an issue applying the generator?
<img width="491" alt="image" src="https://github.com/feldera/feldera/assets/127654/fd8522c4-68b1-4e3d-a25f-0201569c6721">

gz commented:
I think it doesn't just affect arrays but also normal varchar fields are empty

gz reviewed (approved):
Code looks good so we can merge this. But I didn't see any tests, we should have tests for all these cases we encountered during review
